app/src/main/java/com/duckduckgo/app/fire/DataClearerForegroundAppRestartPixel.kt

```diff
@@ -0,0 +1,117 @@
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Stores information about unsent automatic data clearer restart Pixels, detecting if user started the app from an external Intent.
+ * Contains logic to sent unsent pixels.
+ *
+ * When writing values here to SharedPreferences, it is crucial to use `commit = true`. As otherwise the change can be lost in the process restart.
+ */
+@Singleton
+class DataClearerForegroundAppRestartPixel @Inject constructor(
+    private val context: Context,
+    private val pixel: Pixel
+) : LifecycleObserver {
+    private var detectedUserIntent: Boolean = false
+
+    private val pendingAppForegroundRestart: Int
+        get() = preferences.getInt(KEY_UNSENT_CLEAR_APP_RESTARTED_PIXELS, 0)
+
+    private val pendingAppForegroundRestartWithIntent: Int
+        get() = preferences.getInt(KEY_UNSENT_CLEAR_APP_RESTARTED_WITH_INTENT_PIXELS, 0)
+
+    @UiThread
+    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+    fun onAppBackgrounded() {
+        Timber.i("Registered App on_stop")
+        detectedUserIntent = false
+    }
+
+    fun registerIntent(intent: Intent?) {
+        detectedUserIntent = widgetActivity(intent) || !intent?.intentText.isNullOrEmpty()
+    }
+
+    fun incrementCount() {
+        if (detectedUserIntent) {
+            Timber.i("Registered restart with intent")
+            incrementCount(pendingAppForegroundRestart, KEY_UNSENT_CLEAR_APP_RESTARTED_WITH_INTENT_PIXELS)
+        } else {
+            Timber.i("Registered restart w/ intent")
+            incrementCount(pendingAppForegroundRestartWithIntent, KEY_UNSENT_CLEAR_APP_RESTARTED_PIXELS)
+        }
+    }
+
+    fun firePendingPixels() {
+        firePendingPixels(pendingAppForegroundRestart, Pixel.PixelName.FORGET_ALL_AUTO_RESTART)
+        firePendingPixels(pendingAppForegroundRestartWithIntent, Pixel.PixelName.FORGET_ALL_AUTO_RESTART_WITH_INTENT)
+        resetCount()
+    }
+
+    private fun incrementCount(counter: Int, sharedPrefKey: String) {
+        val updated = counter + 1
+        preferences.edit(commit = true) {
+            putInt(sharedPrefKey, updated)
+        }
+    }
+
+    private fun firePendingPixels(counter: Int, pixelName: Pixel.PixelName) {
+        if (counter > 0) {
+            for (i in 1..counter) {
+                Timber.i("Fired pixel: ${pixelName.pixelName}/$counter")
+                pixel.fire(pixelName)
+            }
+        }
+    }
+
+    private fun resetCount() {
+        preferences.edit(commit = true) {
+            putInt(KEY_UNSENT_CLEAR_APP_RESTARTED_PIXELS, 0)
+            putInt(KEY_UNSENT_CLEAR_APP_RESTARTED_WITH_INTENT_PIXELS, 0)
+        }
+        Timber.i("counter reset")
+    }
+
+    private fun widgetActivity(intent: Intent?): Boolean =
+        intent?.component?.className?.contains(SystemSearchActivity::class.java.canonicalName.orEmpty()) == true
+
+    private val preferences: SharedPreferences
```

Comment:
you could make this `lazy` or assign it at start, avoiding multiple calls when referenced